Decomped PlayerStateGroundSpin (100% Matching)

Mars2032 opened the pull request:
PlayerActionGroundMoveControl partially decomped; all fields with types in header, just no names

Mars2032 committed:
decomped PlayerStateGroundSpin

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.cpp:
```diff
@@ -1,5 +1,39 @@
 #include <al/Library/Math/MathUtil.h>
 
 namespace al {
```
clang-format

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.cpp:
```diff
@@ -1,5 +1,39 @@
+        auto vec3 = param_3;
+        auto vec4 = param_4;
+        sead::Vector3f cross;
+        sead::Vector3CalcCommon<f32>::cross(cross, vec3, vec4);
```
```suggestion
        cross.setCross(vec3, vec4);
```
similar below

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.cpp:
```diff
@@ -1,5 +1,39 @@
+        auto vec4 = param_4;
+        sead::Vector3f cross;
+        sead::Vector3CalcCommon<f32>::cross(cross, vec3, vec4);
+        float mag = cross.x * cross.x + cross.y * cross.y + cross.z * cross.z;
```
```suggestion
        float mag = cross.squaredLength();
```

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.cpp:
```diff
@@ -1,5 +1,39 @@
+        sead::Vector3CalcCommon<f32>::cross(cross, vec3, vec4);
+        float mag = cross.x * cross.x + cross.y * cross.y + cross.z * cross.z;
+        if (mag >= 1e-06) {
+            f32 dot = sead::Vector3CalcCommon<f32>::dot(vec2, vec3);
```
```suggestion
            f32 dot = vec2.dot(vec3);
```
similar below

Mars2032 replied:
i removed this little half-assed decomp because it was gross and also didn't really need it right now lol

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.h:
```diff
@@ -37,10 +39,14 @@ void makeMtxRotateTrans(sead::Matrix34f* out, const sead::Vector3f& rotate, cons
 void makeMtxUpFrontPos(sead::Matrix34f* out, const sead::Vector3f& up, const sead::Vector3f& front, const sead::Vector3f& pos);
 
-void separateVectorHV(sead::Vector3<float> *,sead::Vector3<float> *,sead::Vector3<float> const&,sead::Vector3<float> const&);
+void separateVectorHV(sead::Vector3<float>*,sead::Vector3<float>*, sead::Vector3<float> const&, sead::Vector3<float> const&);
```
```suggestion
void separateVectorHV(sead::Vector3f*,sead::Vector3f*, const sead::Vector3f&, const sead::Vector3f&);
```
same below

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <Player/PlayerActionGroundMoveControl.h>
```
```suggestion
#include "Player/PlayerActionGroundMoveControl.h"
```

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.cpp:
```diff
@@ -0,0 +1,22 @@
+
+}
+
+// unfinished
+//void PlayerActionGroundMoveControl::appear() {
+//    auto normal = sead::Vector3f::zero;
+//    rs::calcGroundNormalExceptJustLandOrGravityDir(&normal, mParent, mCollision);
+//    rs::calcGroundNormalOrGravityDir(&mGroundNormal, mParent, mCollision);
+//}
```
Either finish this, or remove from PR - half-done stuff should probably not be added to the repo, if it can be avoided.

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <Player/PlayerActionGroundMoveControl.h>
+#include <math/seadVector.h>
+
+namespace rs {
+    void calcGroundNormalExceptJustLandOrGravityDir(sead::Vector3f*, const al::LiveActor*, const IUsePlayerCollision*);
+    void calcGroundNormalOrGravityDir(sead::Vector3f*, const al::LiveActor*, const IUsePlayerCollision*);
+}
```
Include header instead of forward-declaring functions

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.h:
```diff
@@ -0,0 +1,76 @@
+class PlayerActionTurnControl;
+
+class PlayerActionGroundMoveControl {
+    al::LiveActor* mParent;
```
`private:` missing

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.h:
```diff
@@ -0,0 +1,76 @@
+    f32 mGravityMove;
+    f32 mBrakeSpeed;
+    s32 mCounterBorder;
+    s32 field_0x60;
+    bool field_0x64;
```
```suggestion
    s32 field_60;
    bool field_64;
```

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.h:
```diff
@@ -0,0 +1,76 @@
+
+public:
+    PlayerActionGroundMoveControl(al::LiveActor* parent, const PlayerConst* playerConst, const PlayerInput* input, const IUsePlayerCollision* collision);
+    ~PlayerActionGroundMoveControl() = default;
```
How do you know that this function exists?

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.h:
```diff
@@ -0,0 +1,76 @@
+    void setup(float maxSpeed, float minSpeed, int runFrame, int stickOnBrakeFrame, int brakeFrame, float gravityMove, float , int counterBorder);
+    void updateNormalAndSnap(sead::Vector3f*);
+    void set_field_c4() { field_c4 = true; };
+    sead::Vector3f& getGroundNormal() { return mGroundNormal; };
```
```suggestion
    sead::Vector3f& getGroundNormal() const { return mGroundNormal; };
```

Mars2032 replied:
![image](https://github.com/odyssey-modding/odyssey/assets/90421132/c28ff5ab-c85e-4860-9811-e24b0dadbf92)

MonsterDruide1 commented on src/Player/PlayerActionGroundMoveControl.h:
```diff
@@ -0,0 +1,76 @@
+    void appear();
+    float calcAccelRate(float) const;
+    void calcInitBrakeOnCounter();
+    void setup(float maxSpeed, float minSpeed, int runFrame, int stickOnBrakeFrame, int brakeFrame, float gravityMove, float , int counterBorder);
```
clang-format

MonsterDruide1 commented on src/Player/PlayerAnimator.cpp:
Useless class if it contains no code

MonsterDruide1 commented on src/Player/PlayerAnimator.h:
```diff
@@ -0,0 +1,8 @@
+
+class PlayerAnimator {
+public:
+    void startAnim(sead::SafeString const&);
```
```suggestion
    void startAnim(const sead::SafeString&);
```

MonsterDruide1 commented on src/Player/PlayerInput.cpp:
Useless class

MonsterDruide1 commented on src/Player/PlayerInput.h:
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <math/seadVectorFwd.h>
```
```suggestion
#include <math/seadVector.h>
```

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.h:
```diff
@@ -0,0 +1,27 @@
+class PlayerActionGroundMoveControl;
+
+class PlayerStateGroundSpin : public al::ActorStateBase {
+    const IUsePlayerCollision* mCollision;
```
`private:` missing

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.h:
```diff
@@ -0,0 +1,27 @@
+public:
+    PlayerStateGroundSpin(al::LiveActor* parent, const IUsePlayerCollision* collision,
+                          const PlayerInput* input, const PlayerConst* playerConst, PlayerAnimator* animator);
+    ~PlayerStateGroundSpin() override = default;
```
```suggestion
    ~PlayerStateGroundSpin() override;
```
then add it to the `cpp` to actually generate the function as `=default;`

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.cpp:
```diff
@@ -0,0 +1,63 @@
+                                             PlayerAnimator *animator) : ActorStateBase("地上スピン", parent),
+                                                                         mCollision(collision), mPlayerInput(input), mPlayerConst(playerConst),
+                                                                         mPlayerAnimator(animator) {
+    mIsSpinClockwise = false;
```
Is this line required for matching, or does it also match without it due to it being initialized in the header as well?

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.cpp:
```diff
@@ -0,0 +1,63 @@
+    if (al::isFirstStep(this)) {
+        mPlayerAnimator->startAnim(mIsSpinClockwise ? "SpinGroundR" : "SpinGroundL");
+    }
+    sead::Vector3f someVector = {0.0f, 0.0f, 0.0f};
```
```suggestion
    sead::Vector3f someVector;
```
same below, should give the same assembly

Mars2032 replied:
does not match if i remove the initialization

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.cpp:
```diff
@@ -0,0 +1,63 @@
+    someVector *= mPlayerConst->getGroundSpinBrakeRate();
+    float someVectorLength = someVector.length();
+    float grndSpinMax = mPlayerConst->getGroundSpinMoveSpeedMax();
+    someVectorLength = someVectorLength > grndSpinMax ? someVectorLength : grndSpinMax;
```
```suggestion
    someVectorLength = sead::Mathf::max(someVectorLength, grndSpinMax);
```

Mars2032 replied:
does not match, generates `fcsel s8, s0, s8, mi` instead of `fcsel s8, s8, s0, gt`

MonsterDruide1 replied:
What about flipping the arguments then? (`someVectorLength` second)

Mars2032 replied:
still mismatching, basically the same thing but with the registers flipped (`fcsel s8, s8, s0, mi`) and the instruction before becomes `fcmp s0, s8` rather than `fcmp s8, s0`

Mars2032 committed:
PlayerStateHack and StageScene

Close to matching PlayerStateHack, exeHackDemo is a nightmare though. StageScene has a full header, and everything marked with `unused_<address>` is confirmed to be unused. Also added StageScene dependencies. Wasn't sure where to put some needed `rs` funcs, so decided to forward declare for now.

Mars2032 committed:
PlayerActorHakoniwa

added PlayerActorHakoniwa header and fixed some mistakes in PlayerActorBase. also added ActorDimensionKeeper

Mars2032 committed:
Fixed minor issues, reversed KuriboHack

still some unknowns in KuriboHack but will come back to that potentially at a later date.

MonsterDruide1 commented on data/odyssey_functions.csv:
```diff
@@ -22224,22 +22224,22 @@ Address,Quality,Size,Name
+0x000000710046ecec,O,000212,_ZN21PlayerStateGroundSpinC2EPN2al9LiveActorEPK19IUsePlayerCollisionPK11PlayerInputPK11PlayerConstP14PlayerAnimator
+0x000000710046edc0,O,000068,_ZN21PlayerStateGroundSpin6appearEv
+0x000000710046ee04,O,000576,_ZN21PlayerStateGroundSpin13exeGroundSpinEv
+0x000000710046f044,W,000036,_ZN21PlayerStateGroundSpinD0Ev
+0x000000710046f070,W,000216,_ZN15PlayerStateHackC2EPN2al9LiveActorEP16PlayerHackKeeperP19IPlayerModelChangerP14PlayerAnimatorP7HackCap
```
Do not add `W` states, change them to `M` or `m` depending on how bad they mismatch

Mars2032 replied:
i didn't realize that was even a thing. can you explain the difference? also, was it something i was supposed to edit myself.. ? can do though

MonsterDruide1 replied:
https://github.com/open-ead/nx-decomp-tools/blob/master/viking/src/functions.rs#L23
There are different "states" for functions: Matching (O) is the best, NotDecompiled (U) the most common.
In-between, we have non-matching - major (M) and minor (m) to mark functions which are basically done, but don't match for some reason.
There is also Library (L), but not sure when this would be used.

If you don't manage to get a perfect match, you should manually edit the state to be `M` or `m`, based on how close you are.

MonsterDruide1 commented on lib/al/Library/Camera/CameraUtil.h:
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include <math/seadVectorFwd.h>
```
```suggestion
#include <math/seadVector.h>
```

Mars2032 replied:
whoops, clion auto-include chose the wrong one lmfao

MonsterDruide1 replied:
Unchanged, still wrong

MonsterDruide1 commented on lib/al/Library/Math/MathUtil.cpp:
If you don't change anything in this class, don't touch it

MonsterDruide1 commented on lib/al/Library/Nerve/NerveStateBase.cpp:
```diff
@@ -28,5 +28,5 @@ bool NerveStateBase::update() {
 void NerveStateBase::control() {}
 
-ActorStateBase::ActorStateBase(const char* name, al::LiveActor* actor) : al::NerveStateBase(name), mActor(actor) {}
+ActorStateBase::ActorStateBase(const char* name, al::LiveActor* parent) : al::NerveStateBase(name), mActor(parent) {}
```
Any reason this name seems to be better than `actor`, do you think it captures the idea better?

Why only change it locally then, and not changing the member variable/getter?

Mars2032 replied:
parent is more descriptive i think, so i might rename it everywhere if that's fine to keep it consistent

MonsterDruide1 replied:
Yes, if you think it is a better name, change it everywhere.

MonsterDruide1 commented on lib/al/Library/Scene/SceneHolder.h:
clang-format

MonsterDruide1 commented on lib/al/Library/Scene/SceneInitInfo.h:
clang-format

Mars2032 replied:
not entirely sure what you mean here... running clang-format on this file doesn't change anything for me.

Mars2032 replied:
nevermind i figured it out lol

MonsterDruide1 commented on src/Enemy/KuriboHack.h:
clang-format, should left-align `*` pointers

MonsterDruide1 commented on src/Enemy/KuriboHack.h:
```diff
@@ -0,0 +1,166 @@
+    DisregardReceiver *mDisregardReceiver;
+    u32 mColliderTimer; // really really unsure
+
+public:KuriboHack(const char *);
```
Separate lines

MonsterDruide1 commented on lib/al/Library/Nerve/NerveStateBase.h:
```diff
@@ -11,22 +11,22 @@ class NerveStateBase : public al::NerveExecutor {
 public:
     NerveStateBase(const char* stateName);
-    ~NerveStateBase();
+    inline ~NerveStateBase();
```
Does adding `inline` here change existing matches?

Mars2032 replied:
i didn't try it on everything. it didn't change anything for one of my destructors though, so i'll probably just remove it just to keep everything the same.

fruityloops1 replied:
You should put the function body in the header (also you don't need the inline keyword for classes usually), this is probably what makes the PlayerStateGroundSpin dtor match, I saw the diff when you were working on it and it was calling ~NerveStateBase instead of ~NerveExecutor directly.

MonsterDruide1 commented on src/Layout/StageSceneLayout.h:
```diff
@@ -0,0 +1,84 @@
+} // namespace al
+
+class StageSceneLayout : public al::NerveStateBase {
+    CoinCounter* mCoinCounter;
```
`private:`

MonsterDruide1 commented on src/Layout/StageSceneLayout.h:
```diff
@@ -0,0 +1,84 @@
+    void updateKidsModeLayout();
+    void startShineCountAnim(bool);
+    bool isEndShineCountAnim() const;
+    ~StageSceneLayout();
```
`override`, and right below the `ctor`

MonsterDruide1 commented on src/Player/CapFunction.h:
```diff
@@ -0,0 +1,10 @@
+class HackCap;
+class PlayerAnimator;
+
+namespace CapFunction {
```
Do we know if this is a namespace, or a class with static functions?

Mars2032 replied:
nope. dont know if its possible to know either, same case as GameDataFunction

MonsterDruide1 replied:
According to our current "rule", it should be a class instead.

MonsterDruide1 commented on src/Player/CapFunction.h:
clang-format

MonsterDruide1 commented on src/Player/HackCap.h:
```diff
@@ -0,0 +1,188 @@
+    void syncHackDamageVisibility(bool);
+    void endHack();
+    void startSpinAttack(const char*);
+    void startThrow(bool, const sead::Vector3f&, const sead::Vector3f&,float, const sead::Vector2f&,
```
clang-format

MonsterDruide1 commented on src/Player/HackCap.h:
```diff
@@ -0,0 +1,188 @@
+    void startSpinAttack(const char*);
+    void startThrow(bool, const sead::Vector3f&, const sead::Vector3f&,float, const sead::Vector2f&,
+                    const sead::Vector2f&, const sead::Vector3f&,bool, const sead::Vector3f&,
+                    HackCap::SwingHandType, bool, float, int);
```
```suggestion
                    SwingHandType, bool, float, int);
```

MonsterDruide1 commented on src/Player/HackCap.h:
```diff
@@ -0,0 +1,188 @@
+    void startThrowSeparatePlayJump(const sead::Vector3f&, const sead::Vector3f&, float);
+    void startCatch(const char*, bool, const sead::Vector3f&);
+    void forcePutOn();
+    void forceHack(al::HitSensor*, const CapTargetInfo*); // :eyes:
```
What does this comment mean?

Mars2032 replied:
no idea, i didn't write it lol

MonsterDruide1 commented on src/Player/PlayerActorBase.h:
clang-format

MonsterDruide1 commented on src/Player/PlayerActorHakoniwa.h:
clang-format

MonsterDruide1 commented on src/Player/PlayerAnimator.cpp:
Do not files without code

MonsterDruide1 commented on src/Player/PlayerAnimator.h:
clang-format

MonsterDruide1 commented on src/Player/PlayerHackKeeper.h:
```diff
@@ -1,20 +1,77 @@
+    CapTargetInfo* mStageStartCapTargetInfo;
+    PlayerHackStartTexKeeper* mHackStartTexKeeper;
+    al::LiveActor* mHackModel;
+    sead::PtrArray<void> mHackModelSlices;
```
Put the real type in here, can be done as placeholder-struct

Mars2032 replied:
i have no idea what the type is supposed to be... finding uses of it in ghidra doesn't really help deduce anything at all.

MonsterDruide1 replied:
How did you find a proper name for this entry then?

Mars2032 replied:
gonna be honest i dont think i named this one, it might've been sanae since we share a ghidra project. but the PtrArray is allocated in `PlayerHackKeeper::createHackModel`. nowhere else where its used really helps figure out the type of the array that much.

MonsterDruide1 commented on src/Player/PlayerHackKeeper.h:
```diff
@@ -1,20 +1,77 @@
+    al::LiveActor* mHackModel;
+    sead::PtrArray<void> mHackModelSlices;
+    sead::Matrix34f* field_b8;
+    sead::Vector3i mLastPosition; // unsure
```
Position as `float` seems weird. Can you point out any function that would show this type?

Mars2032 replied:
its just a set of 3 `int`s that appear to be used together, so i thought it might be a vector... but i don't know for sure.

MonsterDruide1 replied:
If it is the set of three `int`s in `createHackModel`, I see nothing related to a position... not sure if my struct is correctly copied though.

Mars2032 replied:
its probably just 3 ints that are loosely related. i'll just change them to 3 ints with no name.

MonsterDruide1 commented on src/Player/PlayerInput.cpp:
Do not add codeless files

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.cpp:
clang-format, should left-align pointers

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateGroundSpin.h:
clang-format

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
clang-format

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    if (mIsStageStartHack) {
+        mIsStageStartHack = false;
+        al::setNerve(this, &nrvPlayerStateHack.Hack);
+        return;
+    }
+    if (mHackKeeper->isPuppetable2()) {
```
```suggestion
    } else if (mHackKeeper->isPuppetable2()) {
```

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    }
+    if (mHackKeeper->isPuppetable2()) {
+        al::setNerve(this, &nrvPlayerStateHack.HackDemoPuppetable);
+        return;
+    }
+    al::setNerve(this, &nrvPlayerStateHack.HackDemo);
```
```suggestion
    } else {
        al::setNerve(this, &nrvPlayerStateHack.HackDemo);
    }
```

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+            reverseCameraFront = -cameraFront;
+
+            al::makeQuatAxisRotation(&quat, downDir, reverseCameraFront, reverseFront, 1.0);
+            sead::Vector3CalcCommon<f32>::rotate(downDir, quat, downDir);
```
This does not exist (yet)?

Mars2032 replied:
correct, it should, but i haven't been able to get a matching impl yet. i do have *an* impl in seadVectorCalcCommon, but its not perfect yet.

MonsterDruide1 replied:
Well, uhh... don't try to PR non-compiling code, please 😅

Mars2032 replied:
it does compile, its just not matching yet.
![image](https://github.com/odyssey-modding/odyssey/assets/90421132/269dc2dc-8129-47d4-a9e2-0cb3b04b5c08)
this is what i've added to seadVectorCalcCommon.hpp, and the function it's used in does still compile, but like i said, its not perfect. it has minor register mismatching still unfortunately.

MonsterDruide1 replied:
What I'm saying is if someone else tries to compile your repo in the current state, it will fail. As long as those changes have not been done to `sead`, this should at least be made a `draft` PR to point out that it's not ready yet.

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    al::setVelocity(getActor(), al::getVelocity(mHackKeeper->getHack()));
+}
+
+void PlayerStateHack::exeHackDemo() {
```
This overall function should get a few cleanup-passthroughs, it's still quite a mess.

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+
+bool PlayerStateHack::isEnableChangeState() const {
+    auto hackKeeper = mHackKeeper;
+    if (hackKeeper->isHack() || (hackKeeper->getHackSensor() && hackKeeper->isStartedHacking()) || al::isNerve(this, &nrvPlayerStateHack.HackDemo)) {
+        return false;
+    } else {
+        return !al::isNerve(this, &nrvPlayerStateHack.HackDemoPuppetable);
+    }
```
Should be equivalent to one giant `return BOOL-ARITHMETIC` block

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+}
+
+bool PlayerStateHack::isEnableCancelHack() const {
+    if (al::isNerve(this, &nrvPlayerStateHack.Hack) && (!mHackKeeper->getHackSensor() || !mHackKeeper->isStartedHacking())) {
+        return al::isGreaterEqualStep(this, 13);
+    }
+    return false;
```
Should be equivalent to one giant `return BOOL-ARITHMETIC` block

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+}
+
+bool PlayerStateHack::isEnableModelSyncShowHide() const {
+    if (al::isNerve(this, &nrvPlayerStateHack.Hack))
+        return mHackKeeper->isHack();
+    else
+        return true;
```
Should be equivalent to one `return BOOL-ARITHMETIC` block

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+}
+
+bool PlayerStateHack::isIgnoreUpdateCollider() const {
+    if (isDead())
+        return true;
+    return !al::isNerve(this, &End);
```
Should be equivalent to one `return BOOL-ARITHMETIC` block

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    f32 sensorDistance = al::calcDistance(source, target);
+    sensorDistance -= 200.0f;
+    sensorDistance = (sensorDistance < 0.0f ? 0.0f : sensorDistance)/30.f;
+    s32 sensorDistInt = (s32)(sensorDistance >= 0 ? sensorDistance + 0.5f : sensorDistance - 0.5f);
```
@ThePixelGamer Does this look like a `sead::MathCalcCommon` function? Like `roundAwayFromZero` or something?

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    al::calcUpDir(&upDir, getActor());
+    f32 sensorDistance = al::calcDistance(source, target);
+    sensorDistance -= 200.0f;
+    sensorDistance = (sensorDistance < 0.0f ? 0.0f : sensorDistance)/30.f;
```
```suggestion
    sensorDistance = sead::Mathf::max(sensorDistance, 0)/30.0f;
```

MonsterDruide1 replied:
This could probably even inline the line above then, as I don't think N would do so many operations on a single local variables in a row... it's probably just a asm/decompiler artifact.

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+    sensorDistance = (sensorDistance < 0.0f ? 0.0f : sensorDistance)/30.f;
+    s32 sensorDistInt = (s32)(sensorDistance >= 0 ? sensorDistance + 0.5f : sensorDistance - 0.5f);
+
+    s32 newSensorDistInt = (sensorDistInt + 15) > 30 ? 30 : sensorDistInt + 15;
```
sead::Mathf::min

MonsterDruide1 commented on src/Player/PlayerState/PlayerStateHack.cpp:
```diff
@@ -0,0 +1,276 @@
+
+    s32 newSensorDistInt = (sensorDistInt + 15) > 30 ? 30 : sensorDistInt + 15;
+
+    mHackDemoStartLength = sensorDistInt < 0 ? 15 : newSensorDistInt;
```
This is weird on its own, should probably also be in-line with the stuff above

MonsterDruide1 commented on src/Scene/StageScene.cpp:
```diff
@@ -0,0 +1,178 @@
+#include <al/Library/Nerve/NerveSetupUtil.h>
+#include <al/Library/Scene/SceneHolder.h>
+#include <al/Library/Nerve/NerveUtil.h>
+#include <System/GameDataFunction.h>
```
```suggestion
#include "System/GameDataFunction.h"
```

MonsterDruide1 commented on src/Scene/StageScene.cpp:
```diff
@@ -0,0 +1,178 @@
+    } nrvStageScene;
+}
+
+namespace rs {
+    void endDemoCommonProc(al::Scene* scene, ProjectItemDirector* projItemDir);
+    bool isPlayerHackRadiconNpc(const al::LiveActor* actor);
+    void setKidsModeLayoutDisable(const al::IUseSceneObjHolder*);
+    void killCoinColletHintEffect(const al::IUseSceneObjHolder*);
+    void updateKitListDemoPlayerWithPauseEffect(al::Scene*);
+    bool isActiveDemo(al::Scene*);
+} // namespace rs
```
no forward-declaring of functions

Mars2032 replied:
i have no idea where to put these functions, so if you could suggest a spot, that would be great. lots of rs functions don't show up in the map properly with a place to put them, or sometimes they don't show up at all. `endDemoCommonProc` has no folder associated with it, and `isPlayerHackRadiconNpc` does not show up in the function map.

MonsterDruide1 replied:
That's a thing for @GRAnimated to fix the function map first.

Mars2032 replied:
... so where do i put them? i have no reference of where rs functions are supposed to go in general.

MonsterDruide1 commented on src/Scene/StageScene.cpp:
```diff
@@ -0,0 +1,178 @@
+    bool isActiveDemo(al::Scene*);
+} // namespace rs
+
+void StageScene::exeDemoGetLifeMaxUpItem() {
```
`if/else-if` block instead of returns

MonsterDruide1 commented on src/Scene/StageScene.cpp:
```diff
@@ -0,0 +1,178 @@
+
+void StageScene::exeDemoHackStart() {
+    auto dataHolder = GameDataHolderAccessor(this);
+    bool isRadicon;
+    if (GameDataFunction::isMainStage(dataHolder)) {
+        isRadicon = false;
+    } else {
+        isRadicon = rs::isPlayerHackRadiconNpc(this->mDemoShine);
+    }
```
```suggestion
    bool isRadicon = !GameDataFunction::isMainStage(dataHolder) && rs::isPlayerHackRadiconNpc(this->mDemoShine);
```

MonsterDruide1 commented on src/Scene/StageScene.h:
```diff
@@ -0,0 +1,265 @@
+class StageSceneStateWarp;
+class StageSceneStateWorldIntroCamera;
+class ScenarioStartCameraHolder;
+class GameDataHolderAccessor;
```
remove this, as it's included

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+#pragma once
+
+#include <basis/seadTypes.h>
```
useless include

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+#pragma once
+
+#include <basis/seadTypes.h>
+#include <math/seadVectorFwd.h>
```
```suggestion
#include <math/seadVector.h>
```

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+#include <basis/seadTypes.h>
+#include <math/seadVectorFwd.h>
+#include <prim/seadSafeString.h>
+#include <nn/g3d/ResFile.h>
```
forward-decl

Mars2032 replied:
what exactly am i forward declaring?

MonsterDruide1 replied:
You should be forward-declaring `ResFile` instead of including the actual source.

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+    GameDataHolder getGameDataHolder(const al::IUseSceneObjHolder*);
+    GameDataHolder getGameDataHolder(const al::SceneObjHolder*);
+    GameDataHolder getGameDataHolder(al::GameDataHolderBase*);
+    GameDataFile getGameDataFile(GameDataHolderWriter);
```
should probably be a pointer

Mars2032 replied:
why exactly? ghidra doesn't show it as a pointer

MonsterDruide1 replied:
I mean the return type.

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+    void tryGetWorldNameByFileId(const al::LayoutActor*, s32);
+    bool isNewSaveDataByFileId(const al::LayoutActor*, s32);
+    s64 getLastUpdateFileTime(const al::LayoutActor*, s32);
+    void makeTextureSaveDataFileName(sead::BufferedSafeStringBase<s8>*, const nn::g3d::ResFile*, const GameDataHolder*, s32);
```
```suggestion
    void makeTextureSaveDataFileName(sead::BufferedSafeString*, const nn::g3d::ResFile*, const GameDataHolder*, s32);
```

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+    class CameraTicket;
+} // namespace al
+
+namespace GameDataFunction {
```
namespace or class?

Mars2032 replied:
doesn't really matter, as long as all functions are made static. namespace looks cleaner IMO, since there's no need see "static" on every single function, but afaik there's no way to know definitively so i chose namespace.

MonsterDruide1 replied:
According to recent discussions, should probably be `class` instead

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,499 @@
+    void resetHintTrans(GameDataHolderAccessor, s32);
+    bool isKoopaLv3(GameDataHolderAccessor);
+
+};
```
```suggestion
}
```

MonsterDruide1 commented on src/System/GameDataHolderAccessor.h:
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include "GameDataHolderWriter.h"
```
absolute include

MonsterDruide1 commented on src/System/GameDataHolderAccessor.h:
```diff
@@ -0,0 +1,15 @@
+class GameDataHolderAccessor : public GameDataHolderWriter
+{
+public:
+    GameDataHolderAccessor(const al::IUseSceneObjHolder* IUseObjHolder);
```
variables should start with lowercase letter

MonsterDruide1 reviewed (changes requested):
Also, you missed a bunch of comments of the last iteration - make sure to click on `load hidden conversations`.

Mars2032 committed:
Fixed numerous things

lots of empty files i added on accident as well as clang-format and general cleaning up of code and fieldnames.

Mars2032 committed:
Minor cleanup

fixed non-matching dtor and cleaned up some PlayerStateHack functions to remove unnessary stack vars

Mars2032 committed:
Minor fixes

sead::Mathf::max change, won't compile until I make a PR to the sead module but I also need to finish Vector3CalcCommon::rotate first to do that lol

Mars2032 committed:
Last Cleanup

i swear this is my last cleanup

MonsterDruide1 commented on lib/al/Library/Nerve/NerveStateBase.h:
```diff
@@ -11,22 +11,23 @@ class NerveStateBase : public al::NerveExecutor {
 public:
     NerveStateBase(const char* stateName);
-    ~NerveStateBase();
+    ~NerveStateBase() {}
```
`override` missing

MonsterDruide1 commented on lib/al/Library/Scene/SceneHolder.h:
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+namespace al {
+class Scene;
```
empty line between forward-decls and function

MonsterDruide1 commented on lib/al/Library/Scene/SceneInitInfo.h:
```diff
@@ -0,0 +1,24 @@
+class AudioDirector;
+
+class SceneInitInfo {
+    const al::GameSystemInfo* mGameSystemInfo;
```
`private` missing

MonsterDruide1 commented on src/Enemy/KuriboHack.h:
```diff
@@ -0,0 +1,166 @@
+}  // namespace al
+
+class KuriboHack : public al::LiveActor {
+public:
```
member variables should be `private` instead

MonsterDruide1 commented on src/Enemy/KuriboHack.h:
```diff
@@ -0,0 +1,166 @@
+namespace al {
+class WaterSurfaceFinder;
+class JointSpringControllerHolder;
+
```
remove empty line here

MonsterDruide1 commented on src/Layout/StageSceneLayout.h:
```diff
@@ -0,0 +1,85 @@
+class SimpleLayoutAppearWaitEnd;
+class GamePadPlayStyleInfo;
+
+namespace al {
+    class LayoutActor;
+    class LayoutInitInfo;
+    class PlayerHolder;
+    class SubCameraRenderer;
+} // namespace al
```
`clang-format`

MonsterDruide1 commented on src/Player/HackFunction.h:
```diff
@@ -0,0 +1,48 @@
+class LiveActor;
+}  // namespace al
+
+namespace HackFunction {
```
`class` instead

MonsterDruide1 commented on src/Player/PlayerActorBase.h:
```diff
@@ -5,17 +5,26 @@
+
 class PlayerActorBase : public al::LiveActor, public IUsePlayerHack {
 private:
-    PlayerHackKeeper* mPlayerHackKeeper;
```
Why are you removing this? I feel like this is not wrong, otherwise you can't return anything useful on `getPlayerHackKeeper()`.

MonsterDruide1 commented on src/Player/PlayerAnimator.h:
```diff
@@ -0,0 +1,80 @@
+class PlayerAnimator {
+public:
+    PlayerAnimator(const PlayerModelHolder*, al::ActorDitherAnimator*);
+    ~PlayerAnimator() = default;
```
Does this function exist, do you see it inlined somewhere else?

MonsterDruide1 commented on src/Player/PlayerInput.h:
```diff
@@ -0,0 +1,144 @@
+class PlayerInput {
+public:
+    PlayerInput(const al::LiveActor*, const IUsePlayerCollision*, const IUseDimension*);
+    ~PlayerInput() = default;
```
Does this exist, do you see it inlined somewhere?

MonsterDruide1 commented on src/Scene/StageScene.cpp:
```diff
@@ -0,0 +1,172 @@
+} nrvStageScene;
+}  // namespace
+
+namespace rs {
+void setKidsModeLayoutDisable(const al::IUseSceneObjHolder*);
+void killCoinColletHintEffect(const al::IUseSceneObjHolder*);
+void updateKitListDemoPlayerWithPauseEffect(al::Scene*);
+bool isActiveDemo(al::Scene*);
+}  // namespace rs
```
Do not forward-declare functions

MonsterDruide1 commented on src/Scene/StageScene.h:
```diff
@@ -0,0 +1,264 @@
+}  // namespace al
+
+class StageScene : public al::Scene {
+public:
```
should be `private`

MonsterDruide1 commented on src/System/GameDataFunction.h:
```diff
@@ -0,0 +1,503 @@
+class SceneObjHolder;
+class GameDataHolderBase;
+class LayoutActor;
+class ActorInitInfo;
+class PlacementInfo;
```
`struct`s

MonsterDruide1 commented on src/System/GameDataHolderAccessor.h:
`clang-format`

MonsterDruide1 commented:
Supersed by https://github.com/MonsterDruide1/OdysseyDecomp/pull/871
